collectors/python.d.plugin/ethtool/ethtool.chart.py

```diff
@@ -0,0 +1,214 @@
+            'dev_{0}_tx_packets_lost'.format(self.id): self.tx_discards_phy,
+        }
+
+
+class Service(ExecutableService):
+    def __init__(self, configuration=None, name=None):
+        ExecutableService.__init__(self, configuration=configuration, name=name)
+        self.dev_filter = self.configuration.get('dev_filter', '')
+        self.order = list()
+        self.definitions = dict()
+        self.s = find_binary('sudo')
+        self.ethtool = find_binary('ethtool')
+        self.ethtool_info = [self.ethtool, '-S']
+        self.devices = self.find_devices()
+        self.metrics = dict()
+        self.max_bw_per_device = self.get_max_bw_per_device()
+
+    def format_data(self, raw_data, name, last_update):
+        rx_bytes = 0
+        rx_discards_phy = 0
+        for line in raw_data:
+            if 'rx_bytes_phy' in line.split(':')[0]:
+                rx_bytes = int(line.split(':')[1])
+            if 'rx_discards_phy' in line.split(':')[0]:
+                rx_discards_phy = int(line.split(':')[1])
+            if 'tx_bytes_phy' in line.split(':')[0]:
+                tx_bytes = int(line.split(':')[1])
+            if 'tx_discards_phy' in line.split(':')[0]:
+                tx_discards_phy = int(line.split(':')[1])
+
+        if name not in self.metrics.keys():
+            self.debug('Init metrics of device {}'.format(name))
+            self.metrics[name] = Metrics(name, 0, 0, 0, 0, 0, 0, 0, 0, last_update)
+            rx_bw = 0
+            rx_bw_util = 0
+            tx_bw = 0
+            tx_bw_util = 0
+        else:
+            nb_sec = (last_update - self.metrics[name].update_time).total_seconds()
+            # self.debug(' rx_bytes {0} last_rx_bytes {1} nb seconds {2}'.format(rx_bytes, self.metrics[name].rx_bytes,
+            #  nb_sec))
+            # Calculation of the bandwidth in Gb/s
+            rx_bw = int((rx_bytes - self.metrics[name].rx_bytes) / nb_sec * 8 / (1000*1000*1000))
+            rx_bw_util = rx_bw / self.max_bw_per_device[name] * 100
+            tx_bw = int((tx_bytes - self.metrics[name].tx_bytes) / nb_sec * 8 / (1000*1000*1000))
+            tx_bw_util = int(tx_bw / self.max_bw_per_device[name] * 100)
+        data = Metrics(name, rx_bytes, rx_bw, rx_bw_util, rx_discards_phy,
+                       tx_bytes, tx_bw, tx_bw_util, tx_discards_phy, last_update)
+        self.debug('toto {}'.format(data.data()))
+        self.metrics[name] = data
+        return data
+
+    def find_devices(self):
+        ls = find_binary('ls')
+        command = [ls, '/sys/class/net']
+        raw_data = self._get_raw_data(command=command)
+        device_names = [s.rstrip() for s in raw_data]
+        devices_filtered = []
+        # Filter devices according to the parameter set in the conf file
+        if self.dev_filter:
+            for device in device_names:
+                if self.dev_filter in device:
+                    devices_filtered.append(device)
```

Comment:
Not exactly critical, but could we extend this to take a regex for the device filter? That would allow us to default to `^eth.*|^en[ops].*`, which should limit things to actual Ethernet devices on a vast majority of systems and would mean that most users would not need to set a custom device filter.

If it’s an issue here we could always just do it in a followup PR instead.